modules/correlation/groupingby.c

```diff
@@ -198,6 +203,11 @@ _clone(LogPipe *s)
   GroupingBy *self = (GroupingBy *) s;
   GroupingBy *cloned;
 
+  if (self->id_counter == NULL)
+    {
+      self->id_counter = id_counter_new();
+    }
```

Comment:
I think the lifecycle of the `id_counter` is more complex than necessary. Let me share some deeper level syslog-ng knowledge.

Managing the lifecycle of a driver's field is unfortunately pretty complex in syslog-ng. The fields can be allocated in `new()`, `init()`, `set_...()`, `clone()`, or sometimes the config parser allocates them for us (`string_list`, ...). They are usually freed in `deinit()` or `free()`. The calling order of these is also complex, a driver can be constructed by the config parser, it can be cloned, syslog-ng can reload, where a new set of drivers are initialized alongside the old ones, syslog-ng can have a failed reload, where the old config's drivers are reinitialized, etc... A driver must be working properly in case of `new()` -> `init()` -> `deinit()` -> `init()` -> `deinit()`, which happens in case of a failed reload. The general rule of thumb is that if it is possible to allocate something in `new()`, do it there, and deallocate it in `free()`, if allocating is only possible in `init()`, do it there and deallocate it in `deinit()`. It is complicated (although possible) to think through the lifecycle if we do not follow this rule of thumb.

I would suggest to allocate the `id_counter` in `new()` (also get the ID there) and `unref()` it in `free()`. In the `clone()` method you can `unref()` the newly created `id_counter` and replace it with `self`'s with a `ref()` (also getting a new ID for the clone there).